0009-api-v4-redesign.md

```diff
@@ -0,0 +1,212 @@
+
+Strangely, there is both a `POST /users/register` endpoint and a `POST /users` endpoint for creating new users. It is not clear to me why that is the case.
+
+# Unresolved questions
+
+The OpenAPI spec provided with this only lists paths, methods, path parameters. There are still several aspects of the API that need to be hashed out:
+
+- Request schemas
+- Response schemas
+- Response codes
+- Query parameters (where appropriate)
+
+The second is especially important since the v3 (and current as of January 12, 2025 v4) API returns very large responses often with more data than what the client needs. I limited the proposal spec to just routes because the changes I propose are already quite large.
+
+One other potential change worth mentioning is adopting [hypertext as the engine of application state (HATEOAS)](https://restfulapi.net/hateoas/) for the API. I don't feel strongly about it myself, but since I framed this proposal as making the API more RESTful and HATEOAS is a component for designing RESTful APIs, it would be remiss of me not to mention it.
+
+Besides all of the above, the design of the routes presented in this proposal's spec should be discussed.
+
+# Future possibilities
+
+One thing that would be very good to have (but not in the scope of this RFC) would be a way to generate an OpenAPI spec from Lemmy's source code. I've found 2 good libraries that can be used for generating an OpenAPI 3.0 spec: [`apistos`](https://crates.io/crates/apistos) and [`oasgen`](https://crates.io/crates/oasgen).
+
+Besides new things, I think taking a closer look at our API and the data it accepts and returns could help with future refactors if any are done. In particular, I think the current codebase lacks a separation of concerns between data used in the database and data returned by the API. The worst offenders in this regard are endpoints that return entire view objects, like [`get_post`](https://github.com/LemmyNet/lemmy/blob/11e05135923b7060df29e4525183314f653b6353/crates/api_crud/src/post/read.rs#L110-L115) which returns both a post view and a community view. A lack of separation of concerns is also visible in the route handler functions, which all handle not only parsing requests and sending responses, but much of the business logic and even some of the database handling logic (the last insofar as `&mut context.pool()` needs to be passed around everywhere).
```

Comment:
> the current codebase lacks a separation of concerns between data used in the database and data returned by the API

Separation of concerns means more code means more maintenance work. It makes sense when you have multiple teams collaborating on a project, to serve as boundaries between the teams. But that is not the case of Lemmy, in our case it makes sense to keep things as simple as possible.

Reply:
In a sense too, our DB views **are** that mapping / separation of concerns. The actual data is in normalized tables like `comment`. `CommentView` is really just selecting all of the specifically useful things that front-ends need, so that they don't have to make a ton of chained requests like: `posts -> tons of comments with ids -> comment creator + comment counts + post + community`

We can always create other versions of those views, but they still need to be built from our source tables anyway.

Reply:
For some reason I interpreted the `*View` types as being more for the database than the API. In retrospect I feel pretty silly not having realized the "view" being referred to was frontend related.